### PR TITLE
logging: Add option to automatically duplicate transient strings

### PR DESCRIPTION
**Pull request by nordic-krch**
Added CONFIG_LOG_AUTO_STRDUP (by default on) which enables scanning of
log message strings in search for %s and performs string duplication
(log_strdup()) if string address is not from strdup buffer pool and
outside read only memory section.
    
Added shell command which can be used to check number of performed
string duplications. It can be used to determine if feature can be
disabled since it degrades performance of the logger (+5us for
scanning 20 characters @64MHz cortex-m4).

This should solve the problem with using %s in the logger messages.

### CODE REVIEW

**Comment by nordic-krch** [comment]
fyi @pdunaj @marc-hb

File: subsys/logging/log_core.c [hunk 1]
```diff
@@ -61,12 +61,113 @@ static k_tid_t proc_tid;
...
+ * @return Mask with %s format specifiers found.
+ */
+static u32_t count_s(const char *str, u32_t nargs)
+{
+	char curr = *str++;
+	bool arm = (curr == '%');
+	u32_t arg = 0;
+	u32_t mask = 0;
+
+	while (curr){
```
Comment (pdunaj) [inline subsys/logging/log_core.c]: `) {`

File: subsys/logging/log_core.c [hunk 2]
```diff
@@ -61,12 +61,113 @@ static k_tid_t proc_tid;
...
+		curr = *str++;
+		if (arm) {
+			if (curr == 's') {
+				mask |= 1 << arg++;
```
Comment (pdunaj) [inline subsys/logging/log_core.c]: no check for overflow

Comment (pdunaj) [reply subsys/logging/log_core.c]: checking for `nargs <= 8 * sizeof(mask)` maybe?

**Comment by pdunaj** [comment]
Hi @nordic-krch , I have put some small comments but I see that this PR is closed now.

**Pull request closed without merging**